### PR TITLE
Slurm-Perf

### PR DESCRIPTION
**Pull request by michael-balint**
- [playbooks/slurm-perf-cluster.yml](https://github.com/michael-balint/deepops/blob/slurm-perf/playbooks/slurm-perf-cluster.yml) for setting up cluster w/ slurm + pyxis + enroot
- [playbooks/slurm-validation.yml](https://github.com/michael-balint/deepops/blob/slurm-perf/playbooks/slurm-validation.yml) for validation using multi-node nccl tests
- [docs/slurm-perf-cluster.md](https://github.com/michael-balint/deepops/blob/slurm-perf/docs/slurm-perf-cluster.md) for documentation

### CODE REVIEW

File: roles/slurm-perf/tasks/packages-redhat.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+---
+- name: yum repos
```
Comment (supertetelman) [inline roles/slurm-perf/tasks/packages-redhat.yml]: Given redhat is not yet a supported OS for this role, do we want to remove this file for now?

File: docs/slurm-perf-cluster.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
+SlurmPerf Deployment Guide
+===
+
+## High-Performance Multi-Node Cluster Deployment Guide
+
+   This guide utilizes a collection of playbooks to configure a cluster, deploy a workload manager, and verify performance.
+
+   It leverages open source tools such as Pyxis and Enroot to optimize Slurm for multi-node Deep Learning jobs beyond the cluster configuration described in the [Slurm Deployment Guide](https://github.com/NVIDIA/deepops/blob/master/docs/slurm-cluster.md). Additional details can be found [here](https://docs.nvidia.com/ngc/multi-node-bert-user-guide).
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: We should not hardcode the repo/branch here for the cluster doc.

File: docs/slurm-perf-cluster.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+   * 1 or more NFS Servers
+   * Optional, 1 Server or VM (PXE provisioning machine)
+   * Internet access
+   * An ssh user that can be used to execute Slurm jobs, has sudo access, and can login to all hosts
+
+## Installation Steps
+
+1. Install a supported operating system on all nodes.
+
+   Install a supported operating system on all servers utilizing the [DGXie](https://github.com/NVIDIA/deepops/blob/master/docs/dgxie-container.md) provisioning container, via a 3rd-party solution (i.e. [MAAS](https://maas.io/), [Foreman](https://www.theforeman.org/)), or server BMC/console.
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: We shouldn't hardcode the dgxie-container url here.

File: docs/slurm-perf-cluster.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+   The `slurm-performance.yml` playbook connects to the login node and executes the NCCL tests against all nodes and GPUs. This checks both the correctness and the performance of the cluster. For a full explanation of what these tests do and what the [results mean](https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md) see the official [NCCL Tests documentation](https://github.com/NVIDIA/nccl-tests).
+
+   This playbook can be run standalone, but is run as the last step in the `slurm-perf-cluster.yml`.
+
+   ```sh
+   # Verify Slurm connectivity across all nodes
+   ansible-playbook playbooks/slurm-validation.yml
+   ```
+
+   If the test is successful you should expect to see output similar tobelow:
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: needs space

File: docs/slurm-perf-cluster.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+   ```sh
+   ansible-playbook playbooks/nfs.yml
+   ```
+
+7. Deploy optimized Slurm software using Ansible
+
+   Run the cluster playbook.
+
+   The `slurm-perf-cluster.yml` playbook bootstraps the cluster, configures NFS, installs/optimizes Slurm, and runs a quick system [validation test](##validation).
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: This anchor link is wrong.

File: docs/slurm-perf-cluster.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
+SlurmPerf Deployment Guide
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: We wanted to get rid of "SlurmPerf Deployment Guide." The top level name of this document should be "High-Performance Multi-Node Cluster Deployment Guide" to be more in-line with what the doc is aiming to acocmplish.

File: roles/slurm-perf/tasks/packages-redhat.yml (hunk 1 above)
Comment (michael-balint) [reply roles/slurm-perf/tasks/packages-redhat.yml]: Yes

File: docs/slurm-perf-cluster.md (hunk 2 above)
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Removed hardcoded references (such as this).

File: docs/slurm-perf-cluster.md (hunk 3 above)
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Removed hardcoded references (such as this).

File: docs/slurm-perf-cluster.md (hunk 4 above)
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Done

File: docs/slurm-perf-cluster.md (hunk 5 above)
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Fixed

File: docs/slurm-perf-cluster.md (hunk 6 above)
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Changed top-level header as suggested

File: docs/slurm-perf-cluster.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+3. Edit the Ansible inventory and group_vars
+
+   Edit the Ansible inventory file and verify connectivity to all nodes.
+
+   Ansible uses an inventory which outlines the servers in the cluster and a set of group variables which playbooks use to customize deployment. The previous step created the `config` directory. It is also advised to copy over the slurm-perf example configuration.
+      
+   ```sh
+   # Copy the slurm-perf config
+   cp -rfp ./config/slurm-perf/ ./config
```
Comment (supertetelman) [inline docs/slurm-perf-cluster.md]: Running this results in a non-zero error code. We should run `cp -rfp ./config/slurm-perf/* ./config` instead.

Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Done

File: docs/slurm-perf-cluster.md [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+## Troubleshooting
+
+### Setup.sh fails due to unsupported ansible version
+
+   DeepOps requires Ansible v2.7.8 or greater. If the setup.sh script fails to achieve this, the latest version of ansible can be installed via pip...
+
+   ```sh
+   # on the provisioning node
+   pip install ansible
```
[the remark this answers is not in the document]
Comment (michael-balint) [reply docs/slurm-perf-cluster.md]: Done

File: config.example/group_vars/all.yml [hunk 9]
```diff
@@ -66,3 +66,14 @@
 
 #software_remove_packages:
 #  - popularity-contest
+
+# users configuration
+# create user `nvidia` with password `deepops` on all nodes
+# run with user-password.yml playbook
```
Comment (dholt) [inline config.example/group_vars/all.yml]: Should be `playbooks/users.yml` vs `playbooks/user-password.yml`, which has a hard-coded user var

Comment (michael-balint) [reply config.example/group_vars/all.yml]: Done

**Pull request merged**